### ISSUE TITLE
Cypher query takes O(2^n) time

### ISSUE THREAD

**Issue by ghost** [neo4j/neo4j#5714]
I'm running neo4j version 2.2.5 on Ubuntu. Cypher queries like this one appear to have O(2^n) runtime, where n is the number of OR clauses:

https://gist.github.com/plt6356/b67be0dde6375966297e

With one OR clause, this query is almost instant. With ten or so OR clauses it takes a few seconds, and with the 25 OR clauses in the Gist it hangs the database, forcing me to hard-kill and manually restart the server process.

**Comment by Mats-SX** [comment]
Hello @plt6356 and thanks for reporting this issue.

We've looked into it, and we've found that this is actually pretty hard to solve. We rely on rewriting our predicates to conjunctive normal form in order to enable index access, and your predicate is written on disjunctive normal form. Rewriting that produces an exponentially growing predicate tree, see for example Wikipedia for reference on this: https://en.wikipedia.org/wiki/Conjunctive_normal_form#Conversion_into_CNF.

So what we'll do is to implement another rewriting limit for these types of predicates, where we'll simply not try to rewrite if the predicate tree is too large. We'll aim this fix towards the 2.3 branch. We'll also investigate whether we can implement some of the smarter techniques for predicate rewriting (the ones mentioned in the Wikipedia article for example), but this will become part of some later minor or major release if it becomes prioritised.
